lgc/elfLinker/GlueShader.cpp

```diff
@@ -0,0 +1,284 @@
+  auto ret = cast<ReturnInst>(fetchFunc->back().getTerminator());
+  Value *result = ret->getOperand(0);
+  BuilderBase builder(ret);
+
+  for (unsigned idx = 0; idx != m_fetches.size(); ++idx) {
+    const auto &fetch = m_fetches[idx];
+    const VertexInputDescription *description = m_fetchDescriptions[idx];
+    unsigned structIdx = idx + m_vsEntryRegInfo.sgprCount + m_vsEntryRegInfo.vgprCount;
+    Type *ty = cast<StructType>(result->getType())->getElementType(structIdx);
+
+    if (description) {
+      // Fetch the vertex.
+      Value *vertex = vertexFetch->fetchVertex(ty, description, fetch.location, fetch.component, builder);
+      result = builder.CreateInsertValue(result, vertex, structIdx);
+    }
+  }
+  ret->setOperand(0, result);
+
+  // Hook up the inputs (vertex buffer, base vertex, base instance, vertex ID, instance ID). The fetchVertex calls
+  // left its uses of them as lgc.special.user.data and lgc.shader.input calls.
+  for (Function &func : *fetchFunc->getParent()) {
+    if (!func.isDeclaration())
+      continue;
+    if (func.getName().startswith(lgcName::SpecialUserData) || func.getName().startswith(lgcName::ShaderInput)) {
+      while (!func.use_empty()) {
+        auto call = cast<CallInst>(func.use_begin()->getUser());
+        Value *replacement = nullptr;
+        switch (cast<ConstantInt>(call->getArgOperand(0))->getZExtValue()) {
+        case static_cast<unsigned>(UserDataMapping::VertexBufferTable): {
+          // Need to extend 32-bit vertex buffer table address to 64 bits.
+          AddressExtender extender(fetchFunc);
+          unsigned highAddr = cast<ConstantInt>(call->getArgOperand(1))->getZExtValue();
+          builder.SetInsertPoint(&*fetchFunc->front().getFirstInsertionPt());
+          replacement = extender.extend(fetchFunc->getArg(m_vsEntryRegInfo.vertexBufferTable), highAddr,
+                                        call->getType(), builder);
+          break;
+        }
+        case static_cast<unsigned>(UserDataMapping::BaseVertex):
+          replacement = fetchFunc->getArg(m_vsEntryRegInfo.baseVertex);
+          break;
+        case static_cast<unsigned>(UserDataMapping::BaseInstance):
+          replacement = fetchFunc->getArg(m_vsEntryRegInfo.baseInstance);
+          break;
+        case static_cast<unsigned>(ShaderInput::VertexId):
+          builder.SetInsertPoint(&*fetchFunc->front().getFirstInsertionPt());
+          replacement = builder.CreateBitCast(fetchFunc->getArg(m_vsEntryRegInfo.sgprCount + m_vsEntryRegInfo.vertexId),
+                                              builder.getInt32Ty());
+          break;
+        case static_cast<unsigned>(ShaderInput::InstanceId):
+          builder.SetInsertPoint(&*fetchFunc->front().getFirstInsertionPt());
+          replacement = builder.CreateBitCast(
+              fetchFunc->getArg(m_vsEntryRegInfo.sgprCount + m_vsEntryRegInfo.instanceId), builder.getInt32Ty());
+          break;
+        default:
```

Comment:
It seems DrawIndex is not listed here?

Reply:
I think a vertex fetch never uses DrawIndex.

Reply:
OK. I didn't notice this is for vertex fetching only.